### PR TITLE
Westend<>Rococo Headers Relay

### PR DESCRIPTION
**Pull request by HCastano**
This PR lays the groundwork for the Rococo<>Westend relay.

~Still in progress since I need to figure out how to make a mock generic Call type for
`submit_finality_proof` (and probably fix some other issues which I haven't encountered
of yet).~

### CODE REVIEW

**Comment by tomusdrw** [comment]
I think it's ready for review, let's get the bridge out :)

File: primitives/chain-rococo/src/lib.rs [hunk 1]
```diff
@@ -42,8 +43,26 @@ pub const VERSION: RuntimeVersion = RuntimeVersion {
 };
 
 #[derive(parity_scale_codec::Encode, parity_scale_codec::Decode, Debug, PartialEq, Eq, Clone)]
+#[allow(non_camel_case_types)]
 pub enum Call {
-	MockPallet,
+	/// Westend bridge pallet.
+	///
+	/// Note this index has to be kept in-sync with Rococo runtime configuration:
+	/// https://github.com/paritytech/polkadot/blob/master/runtime/rococo/src/lib.rs#L272
```
Comment (HCastano) [inline primitives/chain-rococo/src/lib.rs:51]: I don't think this points to what you want it to point to. 

I imagine you want this to point to the Westend bridge pallet instance of Rococo's `construct_runtime!` macro, but it's pointing elsewhere.

We may need to be more explicit in the comment, or tag it to a commit.

File: primitives/chain-rococo/src/lib.rs [hunk 2]
```diff
@@ -42,8 +43,26 @@ pub const VERSION: RuntimeVersion = RuntimeVersion {
...
+	#[codec(index = 40)]
+	BridgeGrandpaWestend(BridgeGrandpaWestendCall),
+}
+
+#[derive(parity_scale_codec::Encode, parity_scale_codec::Decode, Debug, PartialEq, Eq, Clone)]
+#[allow(non_camel_case_types)]
+pub enum BridgeGrandpaWestendCall {
+	#[codec(index = 0)]
```
Comment (HCastano) [inline primitives/chain-rococo/src/lib.rs:59]: We should have a similar comment here. For example, what if somebody flips the extrinsic declarations in the pallet and doesn't update this?

File: primitives/chain-westend/src/lib.rs [hunk 3]
```diff
@@ -43,8 +44,26 @@ pub const VERSION: RuntimeVersion = RuntimeVersion {
 };
 
 #[derive(parity_scale_codec::Encode, parity_scale_codec::Decode, Debug, PartialEq, Eq, Clone)]
+#[allow(non_camel_case_types)]
 pub enum Call {
-	MockModule,
+	/// Rococo bridge pallet.
+	///
+	/// Note this index has to be kept in-sync with westend runtime configuration:
+	/// https://github.com/paritytech/polkadot/blob/master/runtime/westend/src/lib.rs#L809
```
Comment (HCastano) [inline primitives/chain-westend/src/lib.rs:52]: Same thing about the link location

File: primitives/chain-rococo/src/lib.rs [hunk 4]
```diff
@@ -42,8 +43,26 @@ pub const VERSION: RuntimeVersion = RuntimeVersion {
 };
 
 #[derive(parity_scale_codec::Encode, parity_scale_codec::Decode, Debug, PartialEq, Eq, Clone)]
+#[allow(non_camel_case_types)]
 pub enum Call {
```
Comment (HCastano) [inline primitives/chain-rococo/src/lib.rs:47]: We should have a doc comment here with what our goals are for this enum for future readers.

E.g this should be SCALE compatible with what's in the runtime without necessarily using the runtime

**Review by HCastano — COMMENTED** [review]
I can't approve my own PR, but looks good :)

File: relays/bin-substrate/src/rialto_millau/rococo_headers_to_westend.rs [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+/// Rococo-to-Westend finality sync pipeline.
+pub(crate) type RococoFinalityToWestend = SubstrateFinalityToSubstrate<Rococo, Westend, WestendSigningParams>;
+
+impl SubstrateFinalitySyncPipeline for RococoFinalityToWestend {
+	const BEST_FINALIZED_SOURCE_HEADER_ID_AT_TARGET: &'static str = bp_rococo::BEST_FINALIZED_ROCOCO_HEADER_METHOD;
+
+	type TargetChain = Westend;
+
+	fn customize_metrics(params: MetricsParams) -> anyhow::Result<MetricsParams> {
+		Ok(
```
Comment (svyatonik) [inline relays/bin-substrate/src/rialto_millau/rococo_headers_to_westend.rs:38]: I'd either remove them from Westend -> Millau headers relay (would need a change in Grafana/Prometheus config), or from here. Why would we need two samples of "just-for-test" metrics?

File: relays/bin-substrate/src/rialto_millau/westend_headers_to_rococo.rs [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+/// Westend-to-Rococo finality sync pipeline.
+pub(crate) type WestendFinalityToRococo = SubstrateFinalityToSubstrate<Westend, Rococo, RococoSigningParams>;
+
+impl SubstrateFinalitySyncPipeline for WestendFinalityToRococo {
+	const BEST_FINALIZED_SOURCE_HEADER_ID_AT_TARGET: &'static str = bp_westend::BEST_FINALIZED_WESTEND_HEADER_METHOD;
+
+	type TargetChain = Rococo;
+
+	fn customize_metrics(params: MetricsParams) -> anyhow::Result<MetricsParams> {
+		Ok(
```
Comment (svyatonik) [inline relays/bin-substrate/src/rialto_millau/westend_headers_to_rococo.rs:38]: Also please consider removing these metrics from here.

**Review by svyatonik — APPROVED** [review]

File: relays/bin-substrate/src/rialto_millau/rococo_headers_to_westend.rs (hunk 5 above)
Comment (tomusdrw) [reply relays/bin-substrate/src/rialto_millau/rococo_headers_to_westend.rs]: I expect that it's going to be different people looking at these metrics (and they will be visibile in dfferent places).
1. We have bridge-team-internal Grafana and our team looking at Westend<>Millau stuff.
2. We have parity-internal Grafana and our devops team looking at Rococo<>Westend.

Hence I'd keep both. I extracted a function though, so the code is not triplicated.

File: primitives/chain-rococo/src/lib.rs (hunk 4 above)
Comment (tomusdrw) [reply primitives/chain-rococo/src/lib.rs:56]: Added comments and test.

File: primitives/chain-westend/src/lib.rs (hunk 3 above)
Comment (tomusdrw) [reply primitives/chain-westend/src/lib.rs]: Fixed.

File: primitives/chain-rococo/src/lib.rs (hunk 2 above)
Comment (tomusdrw) [reply primitives/chain-rococo/src/lib.rs:65]: Added a test for this.

File: primitives/chain-rococo/src/lib.rs (hunk 1 above)
Comment (tomusdrw) [reply primitives/chain-rococo/src/lib.rs]: Comment updated, please take a look.

**Pull request merged**